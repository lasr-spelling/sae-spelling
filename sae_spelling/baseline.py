import random
from collections.abc import Generator
from dataclasses import dataclass

import torch
from tqdm.autonotebook import tqdm
from transformers import PreTrainedModel, PreTrainedTokenizerFast

from sae_spelling.prompting import create_icl_prompt, spelling_formatter
from sae_spelling.vocab import get_alpha_tokens


@dataclass
class BaselineResult:
    word_length: str
    icl_length: int
    accuracy: float


def get_valid_vocab(
    tokenizer: PreTrainedTokenizerFast, sample_cutoff: int = 1000
) -> dict:
    """
    This function takes in a model tokenizer, and then returns a dictionary of valid tokens to spell from the tokenizer organised by length of token (leading spaces/underscores).

    Args:
    tokenizer (transformers.PreTrainedTokenizer): The tokenizer for the model.
    sample_cutoff (int): Minimum number of tokens that must have the given length for that length to be chosen for testing.

    Outputs:
    dict: A dictionary containing each a token length, and then the collection of tokens that are this long.
    """
    alpha_tokens = get_alpha_tokens(tokenizer)
    total_vocab_dict = {}
    long_token_cutoff = len(
        max(alpha_tokens, key=len)
    )  # find the length of longest token

    for i in range(1, long_token_cutoff + 1):
        _tmplist = [
            s for s in alpha_tokens if len(s.replace("▁", "")) == i
        ]  # don't count the leading underscore for token length (▁cat is spelled the same as cat)
        if len(_tmplist) == 0:
            continue
        total_vocab_dict[str(i)] = _tmplist

    keys_to_delete = [
        i for i in total_vocab_dict.keys() if len(total_vocab_dict[i]) < sample_cutoff
    ]

    for key in keys_to_delete:
        del total_vocab_dict[key]

    return total_vocab_dict


def generate_and_score_samples(
    model: PreTrainedModel,
    tokenizer: PreTrainedTokenizerFast,
    vocab_dict: dict,
    samples_per_combo: int,
    max_icl_length: int = 8,
    capitals: str | None = None,
    char_gap: str = "-",
    example_gap: str = " ",
    batch_size: int = 32,
) -> Generator[BaselineResult]:
    """
    This function takes in various user parameters, iterating over different word lengths and in-context learning (ICL) lengths,
    calculates accuracy scores for each batch and then outputs them to a dict. This can then be turned into a dataframe for plotting and analysis.

    Args:
    model (transformers.PreTrainedModel): The model to use for generation (assumed to be downloaded from Huggingface)
    tokenizer (transformers.PreTrainedTokenizer): The tokenizer for the model.
    vocab_dict (dict): The dictionary of tokens to test for (assumed to be created by the get_valid_vocab function above)
    samples_per_combo (int): Number of samples to generate for each word length x ICL length combination.
    max_icl_length (int, optional): Maximum in-context learning length. Defaults to 8.
    capitals (str, optional): Capitalization style ('upper', 'lower', or None for original case). Defaults to None.
    char_gap (str, optional): Character to use as separator in spellings. Defaults to '-'.
    example_gap (str, optional): Separator between examples in ICL prompts. Defaults to ' '.
    batch_size (int, optional): Batch size for processing. Defaults to 32.

    Yields:
    dict: A dictionary containing 'word_length', 'icl_length', and 'accuracy' for each combination.
    """
    total_combinations = len(vocab_dict) * max_icl_length
    with tqdm(total=total_combinations, desc="Processing combinations") as pbar:
        for word_length in vocab_dict.keys():
            if capitals == "upper":
                words = [
                    v for v in vocab_dict[word_length] if v.replace("▁", "").isupper()
                ]
            elif capitals == "lower":
                words = [
                    v for v in vocab_dict[word_length] if v.replace("▁", "").islower()
                ]
            else:
                words = vocab_dict[word_length]
            tokens_to_gen = (
                int(word_length) * 2 - 1 if char_gap != " " else int(word_length)
            )

            for icl_length in range(1, max_icl_length + 1):
                test_vocab = random.sample(
                    words, k=min(samples_per_combo, len(words) - max_icl_length)
                )
                sample_vocab = [w for w in words if w not in test_vocab]
                all_correct = 0
                total_processed = 0

                for i in range(0, len(test_vocab), batch_size):
                    batch = test_vocab[i : i + batch_size]
                    inputs = []
                    targets = []
                    for w in batch:
                        test_case = create_icl_prompt(
                            w,
                            examples=sample_vocab,
                            example_separator=example_gap,
                            answer_formatter=spelling_formatter(
                                separator=char_gap,
                            ),
                            max_icl_examples=icl_length,
                        )
                        inputs.append(test_case.base)
                        targets.append(test_case.answer)

                    # Process batch
                    with torch.no_grad():
                        input_ids = tokenizer(
                            inputs, padding=True, truncation=True, return_tensors="pt"
                        ).to("cuda")
                        input_length = input_ids["input_ids"].shape[1]  # type: ignore
                        outputs = model.generate(
                            **input_ids,  # type: ignore
                            max_new_tokens=tokens_to_gen,  # type: ignore
                        )
                        answers = tokenizer.batch_decode(
                            outputs[:, input_length:], skip_special_tokens=True
                        )
                    all_correct += sum(a == t for a, t in zip(answers, targets))
                    total_processed += len(batch)
                    torch.cuda.empty_cache()

                accuracy = all_correct / total_processed

                yield BaselineResult(
                    word_length=word_length,
                    icl_length=icl_length,
                    accuracy=accuracy,
                )

                pbar.update(1)
