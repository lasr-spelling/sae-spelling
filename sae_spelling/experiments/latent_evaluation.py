from collections import defaultdict
from pathlib import Path
from typing import Literal

import pandas as pd
import seaborn as sns
import torch
from matplotlib import pyplot as plt
from sae_lens import SAE
from sklearn import metrics
from tqdm.autonotebook import tqdm
from transformers import AutoTokenizer, PreTrainedTokenizerFast
from tueplots import axes, bundles

from sae_spelling.experiments.common import (
    EXPERIMENTS_DIR,
    PROBES_DIR,
    SaeInfo,
    get_gemmascope_saes_info,
    get_or_make_dir,
    humanify_sae_width,
    load_df_or_run,
    load_gemmascope_sae,
    load_probe,
    load_probe_data_split,
)
from sae_spelling.probing import LinearProbe
from sae_spelling.vocab import LETTERS

LATENT_EVALUATION_EXPERIMENT_NAME = "latent_evaluation"
EPS = 1e-8


@torch.inference_mode()
def eval_probe_and_top_sae_raw_scores(
    sae: SAE,
    probe: LinearProbe,
    eval_labels: list[tuple[str, int]],  # list of (token, letter number) pairs
    eval_activations: torch.Tensor,  # n_vocab X d_model
    topk: int = 5,
    metadata: dict[str, str | int | float] = {},
) -> pd.DataFrame:
    norm_probe_weights = probe.weights / torch.norm(probe.weights, dim=-1, keepdim=True)
    norm_W_enc = sae.W_enc / torch.norm(sae.W_enc, dim=0, keepdim=True)
    norm_W_dec = sae.W_dec / torch.norm(sae.W_dec, dim=-1, keepdim=True)
    probe_dec_cos = (norm_probe_weights.to(norm_W_dec.device) @ norm_W_dec.T).cpu()
    probe_enc_cos = (norm_probe_weights.to(norm_W_enc.device) @ norm_W_enc).cpu()
    # Take the topk features by cos sim between the encoder and the probe
    top_sae_feats = probe_enc_cos.topk(topk, dim=-1).indices
    probe = probe.cpu()
    effective_bias = sae.b_enc
    # jumprelu SAEs have a separate threshold which must be passed before a feature can fire
    if hasattr(sae, "threshold"):
        effective_bias = sae.b_enc - sae.threshold
    top_feat_weights = [sae.W_enc.T[top_sae_feats[:, i]].cpu() for i in range(topk)]
    top_feat_biases = [effective_bias[top_sae_feats[:, i]].cpu() for i in range(topk)]

    top_sae_feats_list = top_sae_feats.tolist()

    vocab_scores = []
    for token_act, (token, answer_idx) in tqdm(
        zip(eval_activations, eval_labels), total=len(eval_labels)
    ):
        sae_scores_topk = [
            (token_act @ top_feat_weights[i].T + top_feat_biases[i]).tolist()
            for i in range(topk)
        ]
        probe_scores = probe(token_act).tolist()
        token_scores: dict[str, float | str | int] = {
            "token": token,
            "answer_letter": LETTERS[answer_idx],
            **metadata,
        }
        for letter_i, (letter, probe_score) in enumerate(zip(LETTERS, probe_scores)):
            token_scores[f"score_probe_{letter}"] = probe_score
            for topk_i, sae_scores in enumerate(sae_scores_topk):
                feat_id = int(top_sae_feats_list[letter_i][topk_i])
                sae_score = sae_scores[letter_i]
                token_scores[f"score_sae_{letter}_top_{topk_i}"] = sae_score
                token_scores[f"sae_{letter}_top_{topk_i}_feat"] = feat_id
                token_scores[f"cos_probe_sae_enc_{letter}_top_{topk_i}"] = (
                    probe_enc_cos[letter_i, feat_id].item()
                )
                token_scores[f"cos_probe_sae_dec_{letter}_top_{topk_i}"] = (
                    probe_dec_cos[letter_i, feat_id].item()
                )
        vocab_scores.append(token_scores)
    return pd.DataFrame(vocab_scores)


def build_evaluation_df(results_df, sae_info: SaeInfo, topk: int = 5):
    aucs = []
    for letter in LETTERS:
        y = (results_df["answer_letter"] == letter).values
        pred_probe = results_df[f"score_probe_{letter}"].values
        auc_probe = metrics.roc_auc_score(y, pred_probe)
        f1_probe = metrics.f1_score(y, pred_probe > 0)
        recall_probe = metrics.recall_score(y, pred_probe > 0)
        precision_probe = metrics.precision_score(y, pred_probe > 0)

        auc_info = {
            "auc_probe": auc_probe,
            "f1_probe": f1_probe,
            "recall_probe": recall_probe,
            "precision_probe": precision_probe,
            "letter": letter,
            "sae_l0": sae_info.l0,
            "sae_width": sae_info.width,
        }

        for topk_i in range(topk):
            pred_sae = results_df[f"score_sae_{letter}_top_{topk_i}"].values
            auc_sae = metrics.roc_auc_score(y, pred_sae)
            f1 = metrics.f1_score(y, pred_sae > EPS)
            recall = metrics.recall_score(y, pred_sae > EPS)
            precision = metrics.precision_score(y, pred_sae > EPS)
            auc_info[f"auc_sae_top_{topk_i}"] = auc_sae
            auc_info[f"f1_sae_top_{topk_i}"] = f1
            auc_info[f"recall_sae_top_{topk_i}"] = recall
            auc_info[f"precision_sae_top_{topk_i}"] = precision
            auc_info[f"sae_top_{topk_i}_feat"] = results_df[
                f"sae_{letter}_top_{topk_i}_feat"
            ].values[0]
            auc_info[f"cos_probe_sae_enc_{letter}_top_{topk_i}"] = results_df[
                f"cos_probe_sae_enc_{letter}_top_{topk_i}"
            ].values[0]
            auc_info[f"cos_probe_sae_dec_{letter}_top_{topk_i}"] = results_df[
                f"cos_probe_sae_dec_{letter}_top_{topk_i}"
            ].values[0]
        aucs.append(auc_info)
    return pd.DataFrame(aucs)


@torch.inference_mode()
def load_and_run_eval_probe_and_top_sae_raw_scores(
    sae_info: SaeInfo,
    tokenizer: PreTrainedTokenizerFast,
    probes_dir: Path | str,
) -> pd.DataFrame:
    sae = load_gemmascope_sae(
        layer=sae_info.layer,
        l0=sae_info.l0,
        width=sae_info.width,
    )
    probe = load_probe(layer=sae_info.layer, probes_dir=probes_dir)
    eval_activations, eval_data = load_probe_data_split(
        tokenizer,
        layer=sae_info.layer,
        device="cpu",
        probes_dir=probes_dir,
    )
    df = eval_probe_and_top_sae_raw_scores(
        sae,
        probe,
        eval_data,
        eval_activations,
        metadata={
            "layer": sae_info.layer,
            "sae_l0": sae_info.l0,
            "sae_width": sae_info.width,
        },
    )
    return df


def _consolidate_results_df(
    results: dict[int, list[tuple[pd.DataFrame, SaeInfo]]],
) -> pd.DataFrame:
    auroc_f1_dfs = []
    for layer, result in results.items():
        for auroc_f1_df, sae_info in result:
            auroc_f1_df["layer"] = layer
            auroc_f1_dfs.append(auroc_f1_df)
    df = pd.concat(auroc_f1_dfs)
    df["sae_width_str"] = df["sae_width"].map(humanify_sae_width)
    return df


def plot_metric_vs_l0(
    results: dict[int, list[tuple[pd.DataFrame, SaeInfo]]],
    metric: Literal["f1", "precision", "recall"] = "f1",
    experiment_dir: Path | str = EXPERIMENTS_DIR / LATENT_EVALUATION_EXPERIMENT_NAME,
    layers_range: tuple[int, int] | None = None,
):
    task_output_dir = get_or_make_dir(experiment_dir)
    df = _consolidate_results_df(results)

    title = f"First-letter SAE {metric} vs L0"
    save_file_base = f"{metric}_vs_l0"
    if layers_range is not None:
        df = df[df["layer"].between(*layers_range)]
        title += f" (layers {layers_range[0]}-{layers_range[1] - 1})"
        save_file_base += f"_layers_{layers_range[0]}_{layers_range[1] - 1}"

    sns.set_theme()
    plt.rcParams.update({"figure.dpi": 150})
    with plt.rc_context({**bundles.neurips2021(), **axes.lines()}):
        plt.figure(figsize=(3.75, 2.5))
        sns.scatterplot(
            df[["layer", "sae_l0", "sae_width_str", f"{metric}_sae_top_0"]]
            .groupby(["layer", "sae_width_str", "sae_l0"])  # type: ignore
            .mean()
            .reset_index(),
            x="sae_l0",
            y=f"{metric}_sae_top_0",
            hue="sae_width_str",
            s=15,
            rasterized=True,
        )
        plt.legend(title="SAE width", title_fontsize="small")
        plt.title(title)
        plt.xlabel("L0")
        plt.ylabel(f"Mean {metric}")
        plt.tight_layout()
        plt.savefig(task_output_dir / f"{save_file_base}.pdf")
        plt.show()


def plot_metric_vs_layer(
    results: dict[int, list[tuple[pd.DataFrame, SaeInfo]]],
    metric: Literal["f1", "precision", "recall"] = "f1",
    experiment_dir: Path | str = EXPERIMENTS_DIR / LATENT_EVALUATION_EXPERIMENT_NAME,
):
    task_output_dir = get_or_make_dir(experiment_dir)
    df = _consolidate_results_df(results)

    grouped_df = (
        df[["layer", "sae_l0", "sae_width_str", f"{metric}_sae_top_0"]]
        .groupby(["layer", "sae_width_str", "sae_l0"])
        .mean()
        .reset_index()
    )
    probe_df = df[["layer", "f1_probe"]].groupby(["layer"]).mean().reset_index()

    sns.set_theme()
    plt.rcParams.update({"figure.dpi": 150})
    with plt.rc_context({**bundles.neurips2021(), **axes.lines()}):
        plt.figure(figsize=(3.75, 2.5))
        sns.swarmplot(
            data=grouped_df,
            x="layer",
            y=f"{metric}_sae_top_0",
            hue="sae_width_str",
            size=2,
        )

        # Add the line plot for probe_df
        sns.lineplot(
            data=probe_df,
            x="layer",
            y="f1_probe",
            color="gray",
            linewidth=1,
            marker="o",
            markersize=3,
            label="Probe",
        )

        current_ticks = plt.gca().get_xticks()
        if len(current_ticks) > 20:
            # Select every other tick
            current_labels = [int(tick) for tick in current_ticks]
            new_ticks = current_ticks[::2]
            new_labels = current_labels[::2]
            plt.xticks(new_ticks, new_labels)  # type: ignore

        # Customize the plot
        plt.title(f"First-letter SAE {metric} vs Layer")
        plt.xlabel("Layer")
        plt.ylabel(f"Mean {metric}")
        plt.legend(
            title="SAE width",
            bbox_to_anchor=(1.05, 1),
            loc="upper left",
            title_fontsize="small",
        )

        # Adjust layout to prevent clipping of the legend
        plt.tight_layout()
        plt.savefig(task_output_dir / f"{metric}_vs_layer.pdf")
        plt.show()


def run_latent_evaluation_experiments(
    layers: list[int],
    experiment_dir: Path | str = EXPERIMENTS_DIR / LATENT_EVALUATION_EXPERIMENT_NAME,
    probes_dir: Path | str = PROBES_DIR,
    force: bool = False,
    skip_1m_saes: bool = True,
    skip_32k_saes: bool = True,
    skip_262k_saes: bool = True,
    skip_524k_saes: bool = True,
) -> dict[int, list[tuple[pd.DataFrame, SaeInfo]]]:
    task_output_dir = get_or_make_dir(experiment_dir)

    results_by_layer: dict[int, list[tuple[pd.DataFrame, SaeInfo]]] = defaultdict(list)
    tokenizer: PreTrainedTokenizerFast = AutoTokenizer.from_pretrained(
        "google/gemma-2-2b"
    )  # type: ignore
    with tqdm(total=len(layers)) as pbar:
        for layer in layers:
            pbar.set_description(f"Layer {layer}")
            results_by_layer[layer] = []
            sae_infos = get_gemmascope_saes_info(layer)
            for sae_info in sae_infos:
                if skip_1m_saes and sae_info.width == 1_000_000:
                    continue
                if skip_32k_saes and sae_info.width == 32_000:
                    continue
                if skip_262k_saes and sae_info.width == 262_000:
                    continue
                if skip_524k_saes and sae_info.width == 524_000:
                    continue
                raw_results_path = (
                    task_output_dir
                    / f"layer_{layer}_{sae_info.width}_{sae_info.l0}_raw_results.parquet"
                )
                auroc_results_path = (
                    task_output_dir
                    / f"layer_{layer}_{sae_info.width}_{sae_info.l0}_auroc_f1.parquet"
                )

                def get_raw_results_df():
                    return load_df_or_run(
                        lambda: load_and_run_eval_probe_and_top_sae_raw_scores(
                            sae_info, tokenizer, probes_dir
                        ),
                        raw_results_path,
                        force=force,
                    )

                auroc_results_df = load_df_or_run(
                    lambda: build_evaluation_df(get_raw_results_df(), sae_info),
                    auroc_results_path,
                    force=force,
                )
                results_by_layer[layer].append((auroc_results_df, sae_info))
            pbar.update(1)
    return results_by_layer
