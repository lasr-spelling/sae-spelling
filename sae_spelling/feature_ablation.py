from collections.abc import Callable
from dataclasses import dataclass
from typing import Sequence

import torch
from sae_lens import SAE
from transformer_lens import HookedTransformer
from transformer_lens.hook_points import HookPoint

from sae_spelling.sae_utils import (
    SaeReconstructionCache,
    apply_saes_and_run,
)
from sae_spelling.util import batchify


@dataclass
class FeatureAblationsOutput:
    sae_cache: SaeReconstructionCache
    ablation_scores: dict[int, float]
    original_score: float


@torch.no_grad()
def calculate_individual_feature_ablations(
    model: HookedTransformer,
    input: str,
    metric_fn: Callable[[torch.Tensor], torch.Tensor],
    sae: SAE,
    ablate_token_index: int,  # TODO: this can be extended to multiple tokens
    ablate_features: Sequence[int] | None = None,
    return_logits: bool = True,
    batch_size: int = 10,
    show_progress: bool = True,
    # TODO: support not including the error term
) -> FeatureAblationsOutput:
    """
    Calculate the effect of ablating each feature individually. This means return the change in metric if the
    SAE feature never fired at all during the forward pass on the input.
    This is what feature attribution is approximating, but the signs are flipped since attribution asks "How important is a feature?",
    while ablation asks "what is the metric delta if the feature is deleted?".
    If we're unsure about how accurate attribution is, we can compare it to this function (after flipping the sign). This function is much slower, however.

    Args:
        model: The model to run the forward pass on.
        input: The input to the model.
        metric_fn: A function that takes the model output and returns a scalar metric.
        sae: The SAE to use for ablation.
        ablate_features: The indices of the features to ablate. If empty, ablate all firing features.
        ablate_token_index: The index of the token to ablate.
        return_logits: Whether to return the logits or the loss.
        batch_size: The batch size to use for ablation.
        show_progress: Whether to show a progress bar.
    """
    input_toks = model.to_tokens(input)
    hook_point = sae.cfg.hook_name
    original_output = apply_saes_and_run(
        model,
        saes={hook_point: sae},
        input=input_toks,
        return_type="logits" if return_logits else "loss",
        include_error_term=True,
    )
    original_score = metric_fn(original_output.model_output).item()
    if ablate_features is None:
        sae_acts = original_output.sae_activations[hook_point]
        ablate_features = (
            torch.nonzero(sae_acts.feature_acts[0, ablate_token_index])
            .squeeze(-1)
            .tolist()
        )
    ablation_scores = {}
    for batch in batchify(ablate_features, batch_size, show_progress=show_progress):
        feature_vals = original_output.sae_activations[hook_point].feature_acts[
            0, ablate_token_index, batch
        ]
        batch_deltas = -1 * feature_vals.unsqueeze(-1) * sae.W_dec[batch]

        def ablation_hook(value: torch.Tensor, hook: HookPoint):  # noqa: ARG001
            value[:, ablate_token_index, :] += batch_deltas
            return value

        outputs = model.run_with_hooks(
            input_toks.repeat(len(batch), 1),
            fwd_hooks=[(hook_point, ablation_hook)],
        )
        for feat_idx, output in zip(batch, outputs):
            score = metric_fn(output.unsqueeze(0)).item()
            ablation_scores[feat_idx] = score - original_score
    return FeatureAblationsOutput(
        sae_cache=original_output.sae_activations[hook_point],
        ablation_scores=ablation_scores,
        original_score=original_score,
    )
