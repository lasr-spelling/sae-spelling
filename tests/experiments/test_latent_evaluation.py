import torch
from sae_lens import SAE
from syrupy.assertion import SnapshotAssertion

from sae_spelling.experiments.latent_evaluation import (
    eval_probe_and_top_sae_raw_scores,
)
from sae_spelling.probing import LinearProbe


@torch.no_grad()
def test_eval_probe_and_top_sae_raw_scores_gives_sane_results(
    gpt2_l4_sae: SAE, snapshot: SnapshotAssertion
):
    fake_probe = LinearProbe(768, 26)
    fake_probe.weights[2, :] = gpt2_l4_sae.W_enc[:, 123]  # set C to feature 123

    eval_data = [("dog", 3), ("cat", 2), ("fish", 5), ("bird", 1)]
    eval_activations = torch.randn(4, 768)
    df = eval_probe_and_top_sae_raw_scores(
        gpt2_l4_sae,
        fake_probe,
        eval_data,
        eval_activations,
        metadata={"layer": 4},
        topk=2,
    )
    assert all(df["sae_c_top_0_feat"] == 123)
    assert df.columns.values.tolist() == snapshot
