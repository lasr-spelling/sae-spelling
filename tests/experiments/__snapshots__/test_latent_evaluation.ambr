# serializer version: 1
# name: test_eval_probe_and_top_sae_raw_scores_gives_sane_results
  list([
    'token',
    'answer_letter',
    'layer',
    'score_probe_a',
    'score_sae_a_top_0',
    'sae_a_top_0_feat',
    'cos_probe_sae_enc_a_top_0',
    'cos_probe_sae_dec_a_top_0',
    'score_sae_a_top_1',
    'sae_a_top_1_feat',
    'cos_probe_sae_enc_a_top_1',
    'cos_probe_sae_dec_a_top_1',
    'score_probe_b',
    'score_sae_b_top_0',
    'sae_b_top_0_feat',
    'cos_probe_sae_enc_b_top_0',
    'cos_probe_sae_dec_b_top_0',
    'score_sae_b_top_1',
    'sae_b_top_1_feat',
    'cos_probe_sae_enc_b_top_1',
    'cos_probe_sae_dec_b_top_1',
    'score_probe_c',
    'score_sae_c_top_0',
    'sae_c_top_0_feat',
    'cos_probe_sae_enc_c_top_0',
    'cos_probe_sae_dec_c_top_0',
    'score_sae_c_top_1',
    'sae_c_top_1_feat',
    'cos_probe_sae_enc_c_top_1',
    'cos_probe_sae_dec_c_top_1',
    'score_probe_d',
    'score_sae_d_top_0',
    'sae_d_top_0_feat',
    'cos_probe_sae_enc_d_top_0',
    'cos_probe_sae_dec_d_top_0',
    'score_sae_d_top_1',
    'sae_d_top_1_feat',
    'cos_probe_sae_enc_d_top_1',
    'cos_probe_sae_dec_d_top_1',
    'score_probe_e',
    'score_sae_e_top_0',
    'sae_e_top_0_feat',
    'cos_probe_sae_enc_e_top_0',
    'cos_probe_sae_dec_e_top_0',
    'score_sae_e_top_1',
    'sae_e_top_1_feat',
    'cos_probe_sae_enc_e_top_1',
    'cos_probe_sae_dec_e_top_1',
    'score_probe_f',
    'score_sae_f_top_0',
    'sae_f_top_0_feat',
    'cos_probe_sae_enc_f_top_0',
    'cos_probe_sae_dec_f_top_0',
    'score_sae_f_top_1',
    'sae_f_top_1_feat',
    'cos_probe_sae_enc_f_top_1',
    'cos_probe_sae_dec_f_top_1',
    'score_probe_g',
    'score_sae_g_top_0',
    'sae_g_top_0_feat',
    'cos_probe_sae_enc_g_top_0',
    'cos_probe_sae_dec_g_top_0',
    'score_sae_g_top_1',
    'sae_g_top_1_feat',
    'cos_probe_sae_enc_g_top_1',
    'cos_probe_sae_dec_g_top_1',
    'score_probe_h',
    'score_sae_h_top_0',
    'sae_h_top_0_feat',
    'cos_probe_sae_enc_h_top_0',
    'cos_probe_sae_dec_h_top_0',
    'score_sae_h_top_1',
    'sae_h_top_1_feat',
    'cos_probe_sae_enc_h_top_1',
    'cos_probe_sae_dec_h_top_1',
    'score_probe_i',
    'score_sae_i_top_0',
    'sae_i_top_0_feat',
    'cos_probe_sae_enc_i_top_0',
    'cos_probe_sae_dec_i_top_0',
    'score_sae_i_top_1',
    'sae_i_top_1_feat',
    'cos_probe_sae_enc_i_top_1',
    'cos_probe_sae_dec_i_top_1',
    'score_probe_j',
    'score_sae_j_top_0',
    'sae_j_top_0_feat',
    'cos_probe_sae_enc_j_top_0',
    'cos_probe_sae_dec_j_top_0',
    'score_sae_j_top_1',
    'sae_j_top_1_feat',
    'cos_probe_sae_enc_j_top_1',
    'cos_probe_sae_dec_j_top_1',
    'score_probe_k',
    'score_sae_k_top_0',
    'sae_k_top_0_feat',
    'cos_probe_sae_enc_k_top_0',
    'cos_probe_sae_dec_k_top_0',
    'score_sae_k_top_1',
    'sae_k_top_1_feat',
    'cos_probe_sae_enc_k_top_1',
    'cos_probe_sae_dec_k_top_1',
    'score_probe_l',
    'score_sae_l_top_0',
    'sae_l_top_0_feat',
    'cos_probe_sae_enc_l_top_0',
    'cos_probe_sae_dec_l_top_0',
    'score_sae_l_top_1',
    'sae_l_top_1_feat',
    'cos_probe_sae_enc_l_top_1',
    'cos_probe_sae_dec_l_top_1',
    'score_probe_m',
    'score_sae_m_top_0',
    'sae_m_top_0_feat',
    'cos_probe_sae_enc_m_top_0',
    'cos_probe_sae_dec_m_top_0',
    'score_sae_m_top_1',
    'sae_m_top_1_feat',
    'cos_probe_sae_enc_m_top_1',
    'cos_probe_sae_dec_m_top_1',
    'score_probe_n',
    'score_sae_n_top_0',
    'sae_n_top_0_feat',
    'cos_probe_sae_enc_n_top_0',
    'cos_probe_sae_dec_n_top_0',
    'score_sae_n_top_1',
    'sae_n_top_1_feat',
    'cos_probe_sae_enc_n_top_1',
    'cos_probe_sae_dec_n_top_1',
    'score_probe_o',
    'score_sae_o_top_0',
    'sae_o_top_0_feat',
    'cos_probe_sae_enc_o_top_0',
    'cos_probe_sae_dec_o_top_0',
    'score_sae_o_top_1',
    'sae_o_top_1_feat',
    'cos_probe_sae_enc_o_top_1',
    'cos_probe_sae_dec_o_top_1',
    'score_probe_p',
    'score_sae_p_top_0',
    'sae_p_top_0_feat',
    'cos_probe_sae_enc_p_top_0',
    'cos_probe_sae_dec_p_top_0',
    'score_sae_p_top_1',
    'sae_p_top_1_feat',
    'cos_probe_sae_enc_p_top_1',
    'cos_probe_sae_dec_p_top_1',
    'score_probe_q',
    'score_sae_q_top_0',
    'sae_q_top_0_feat',
    'cos_probe_sae_enc_q_top_0',
    'cos_probe_sae_dec_q_top_0',
    'score_sae_q_top_1',
    'sae_q_top_1_feat',
    'cos_probe_sae_enc_q_top_1',
    'cos_probe_sae_dec_q_top_1',
    'score_probe_r',
    'score_sae_r_top_0',
    'sae_r_top_0_feat',
    'cos_probe_sae_enc_r_top_0',
    'cos_probe_sae_dec_r_top_0',
    'score_sae_r_top_1',
    'sae_r_top_1_feat',
    'cos_probe_sae_enc_r_top_1',
    'cos_probe_sae_dec_r_top_1',
    'score_probe_s',
    'score_sae_s_top_0',
    'sae_s_top_0_feat',
    'cos_probe_sae_enc_s_top_0',
    'cos_probe_sae_dec_s_top_0',
    'score_sae_s_top_1',
    'sae_s_top_1_feat',
    'cos_probe_sae_enc_s_top_1',
    'cos_probe_sae_dec_s_top_1',
    'score_probe_t',
    'score_sae_t_top_0',
    'sae_t_top_0_feat',
    'cos_probe_sae_enc_t_top_0',
    'cos_probe_sae_dec_t_top_0',
    'score_sae_t_top_1',
    'sae_t_top_1_feat',
    'cos_probe_sae_enc_t_top_1',
    'cos_probe_sae_dec_t_top_1',
    'score_probe_u',
    'score_sae_u_top_0',
    'sae_u_top_0_feat',
    'cos_probe_sae_enc_u_top_0',
    'cos_probe_sae_dec_u_top_0',
    'score_sae_u_top_1',
    'sae_u_top_1_feat',
    'cos_probe_sae_enc_u_top_1',
    'cos_probe_sae_dec_u_top_1',
    'score_probe_v',
    'score_sae_v_top_0',
    'sae_v_top_0_feat',
    'cos_probe_sae_enc_v_top_0',
    'cos_probe_sae_dec_v_top_0',
    'score_sae_v_top_1',
    'sae_v_top_1_feat',
    'cos_probe_sae_enc_v_top_1',
    'cos_probe_sae_dec_v_top_1',
    'score_probe_w',
    'score_sae_w_top_0',
    'sae_w_top_0_feat',
    'cos_probe_sae_enc_w_top_0',
    'cos_probe_sae_dec_w_top_0',
    'score_sae_w_top_1',
    'sae_w_top_1_feat',
    'cos_probe_sae_enc_w_top_1',
    'cos_probe_sae_dec_w_top_1',
    'score_probe_x',
    'score_sae_x_top_0',
    'sae_x_top_0_feat',
    'cos_probe_sae_enc_x_top_0',
    'cos_probe_sae_dec_x_top_0',
    'score_sae_x_top_1',
    'sae_x_top_1_feat',
    'cos_probe_sae_enc_x_top_1',
    'cos_probe_sae_dec_x_top_1',
    'score_probe_y',
    'score_sae_y_top_0',
    'sae_y_top_0_feat',
    'cos_probe_sae_enc_y_top_0',
    'cos_probe_sae_dec_y_top_0',
    'score_sae_y_top_1',
    'sae_y_top_1_feat',
    'cos_probe_sae_enc_y_top_1',
    'cos_probe_sae_dec_y_top_1',
    'score_probe_z',
    'score_sae_z_top_0',
    'sae_z_top_0_feat',
    'cos_probe_sae_enc_z_top_0',
    'cos_probe_sae_dec_z_top_0',
    'score_sae_z_top_1',
    'sae_z_top_1_feat',
    'cos_probe_sae_enc_z_top_1',
    'cos_probe_sae_dec_z_top_1',
  ])
# ---
